race-buffer/src/reader.rs

```diff
@@ -0,0 +1,139 @@
+use crate::{get_cursor_index, Entry};
+use core::marker::PhantomData;
+use std::cmp::min;
+use std::error::Error;
+
+type Result<T> = std::result::Result<T, Box<dyn Error>>;
+/// Used to "snap" observations of the RaceBuffer's write cursor
+/// or entries in the RaceBuffer's backing storage. Each Snapper access
+/// should have sequential consistency with the memory access of the RaceBuffer writer
+pub trait Snapper<E>
+where
+    E: Entry,
+{
+    /// Take a snapshot of the write cursor of the RaceBuffer
+    fn snap_wcurs(&self) -> Result<usize>;
+    /// Take a snapshot of the overwrite cursor of the RaceBuffer
+    fn snap_owcurs(&self) -> Result<usize>;
+    /// Take a snapshot of the RaceBuffer's backing storage at the given index
+    fn snap_storage(&self, index: usize) -> Result<E>;
+}
+
+/// Struct used to read from a RaceBuffer asynchronously
+pub struct RaceBufferReader<E, S>
+where
+    E: Entry,
+    S: Snapper<E>,
+{
+    /// Struct for reading writer state
+    snapper: S,
+    /// Reader's cursor
+    rcurs: usize,
+    /// Capacity of backing storage
+    storage_cap: usize,
+    /// Cached prefix, not put into buffer until suffix is successfully read
+    stored_prefix: Option<E>,
+    phantom: PhantomData<E>,
+}
+
+impl<E, S> RaceBufferReader<E, S>
+where
+    E: Copy + PartialEq + Entry,
+    S: Snapper<E>,
+{
+    /// Create new RaceBufferReader
+    pub fn new(snapper: S, storage_cap: usize) -> RaceBufferReader<E, S> {
+        RaceBufferReader {
+            snapper,
+            rcurs: 0,
+            storage_cap,
+            stored_prefix: None,
+            phantom: PhantomData,
+        }
+    }
+
+    /// Attempt to read all new entries in race buffer into given read buffer
+    pub fn read(&mut self, rbuf: &mut Vec<Option<E>>) -> Result<()> {
```

Comment:
A possible API adjustment (not critical) that would save some space and maybe be a bit clearer to interpret:

Rather than handling possibly-missing items in the out-buffer with `Option`s like `vec![None, None, Some(e), Some(e)]`, how about instead storing the direct entries and returning the number missed?

E.G. `fn read(&mut self, rbuf: &mut Vec<E>) -> Result<usize, S::Error>`

This could save users of the library downstream Option-unwrapping (and, less importantly, reduce the memory footprint of the needed rbuf).